### PR TITLE
refactor(api): significant reworking of XorUrlEncoder

### PR DESCRIPTION
**Pull request by dan-da**
Overall goals of this PR are to:
1. Unify and simplify usage of urls in nrs and xor format for API users.
2. Make nrs urls easier to work with.
3. Improve validation/correctness
4. Provide easy get/set for all parts of a standard URL, esp query string params.
5. synchronize content_version property and ?v query string param.
6. Perform proper percent decoding/encoding of path and query string
7. Improve documentation/comments.
8. Begin name change: XorUrlEncoder --> SafeUrl
9. Create tests for all of the above.

Changes to XorUrlEncoder:
* adds support for nrs urls
* percent encoding for paths
* improved validation
* to_string() no longer returns Result
* get/set individual url query params
* begin rename of XorUrlEncoder to SafeUrl, with XorUrlEncoder alias for now.
* add many test cases
* comment all public methods
* keep content_version and ?v=1 in query_string in sync
* ::from_url() supports both xorurl and nrsurl
* add ::scheme() getter
* add ::xorurl_host() getter
* add ::nrs_host() getter
* add ::host() getter
* add ::tld() getter
* add ::path_decoded() getter
* add ::query_string_with_separator() getter
* add ::query_pairs() getter
* add ::query_key() getter
* add ::query_key_last() getter
* add ::query_key_first() getter
* add ::fragment_with_separator() getter
* add ::is_nrs() boolean getter
* add ::to_xorurl_string()
* add ::to_nrsurl_string()
* add ::host_to_base()

Additional Changes:
* fixes urls with percent-encoded paths in 'files tree' and 'files ls'
* simplifies XorUrlEncoder::to_string() for various callers, no unwrap needed.
* add SafeUrlParts, replaces helpers::extract_all_url_parts
* add a few extra fields to 'safe xorurl decode' output

Note: if this were a total clean room design, I would do things a bit differently.   As it is, I made things about as clean/correct as I could without changing the present interface too drastically, which would require changing a lot of callers.

Once this PR is merged, I plan to create a second PR changing all references XorUrlEncoder to SafeUrl within safe-api crate, but I wanted to keep this PR smaller and easier to review, so for now I just made an alias XorUrlEncoder --> SafeUrl.

### CODE REVIEW

**Comment by ravinderjangra** [comment]
Great work @dan-da. The changes simplify the implementation.

I didn't understand the use of `SafeUrlParts` completely but I'll take a look again later.

File: safe-api/src/api/app/xorurl.rs [hunk 1]
```diff
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
...
-    pub fn to_string(&self) -> Result<String> {
-        self.to_base(DEFAULT_XORURL_BASE)
+
+    /// serializes the URL to an XorUrl string.
+    ///
+    /// This function may be called on an NrsUrl and
+    /// the corresponding XorUrl will be returned.
+    pub fn to_xorurl_string(&self) -> String {
+        self.to_base(DEFAULT_XORURL_BASE).unwrap_or_else(|e| {
+            warn!("{}", e);
```
Comment (bochaco) [inline safe-api/src/api/app/xorurl.rs:843]: I don't think this is a desirable way of not returning an error, we are preventing the consumer from knowing what happened, when I call this function, since it doesn't return a Result<>, I then assume it will never fail but it can fail and it doesn't even tell me so, using an empty string as the way to check errors is going against the language way of handling errors properly.

File: safe-api/src/api/app/xorurl.rs [hunk 2]
```diff
@@ -547,12 +1073,163 @@ impl XorUrlEncoder {
             base,
         )
     }
+
+    // utility to generate a dummy url from a query string.
+    fn query_string_to_url(query_str: &str) -> Result<Url> {
```
Comment (bochaco) [inline safe-api/src/api/app/xorurl.rs:1078]: Assuming this is for testing it should go under `mod tests {`

File: safe-api/src/api/app/xorurl.rs [hunk 3]
```diff
@@ -561,10 +1238,118 @@ mod tests {
...
+            None,
+            None,
+            None,
+            None,
+            None,
+        )
+        .expect_err(msg);
+        match result {
+            Error::InvalidMediaType(e) => assert!(e.contains("You can use 'SafeContentType::Raw'")),
+            _ => panic!(wrong_err),
```
Comment (bochaco) [inline safe-api/src/api/app/xorurl.rs:1265]: We should not panic on tests failures, but return an Error, see in other tests we use `Error::Unexpected` error type

File: safe-api/src/api/app/xorurl.rs [hunk 4]
```diff
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
...
+            } else {
+                pairs.append_pair(&k, &v);
+            }
+        }
+        if !set_key {
+            if let Some(v) = val {
+                pairs.append_pair(key, v);
+            }
+        }
+        std::mem::drop(pairs);
```
Comment (bochaco) [inline safe-api/src/api/app/xorurl.rs:726]: Do we really need this call?

**Comment by dan-da** [comment]
> I didn't understand the use of `SafeUrlParts` completely but I'll take a look again later.

Previously there was a helper function extract_all_url_parts() that returned sub_domains, domain (tld), path, query_string, fragment, version in a lengthy tuple.  It was a fairly thin wrapper around Url::parse() with some validation.  I thought it would be cleaner to make a real struct instead, with all the standard url parts including scheme, host, and move the extract/parse func into it.  Version parsing moved into SafeUrl/XorUrlEncoder.

So basically SafeUrlParts is useful if you just want to parse a SafeUrl and get the standard url parts, but without performing any xor decoding.  It is presently internal to the API, not for public consumption, and is used by SafeUrl.

File: safe-api/src/api/app/xorurl.rs [hunk 5]
```diff
@@ -148,32 +160,164 @@ impl SafeDataType {
...
+///
+// if nrs_host is non-empty, it is considered an nrsurl.
+// else an xorurl.  see: SafeUrl::is_nrs()
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
-pub struct XorUrlEncoder {
+pub struct SafeUrl {
     encoding_version: u64, // currently only v1 supported
-    xorname: XorName,
+    xorname: XorName,      // applies to nrsurl and xorurl
+    nrs_host: String,      // full hostname, only for nrsurl
```
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:265]: ok, maybe nrs_name.

File: safe-api/src/api/app/xorurl.rs [hunk 6]
```diff
@@ -148,32 +160,164 @@ impl SafeDataType {
...
+    /// * `data_type` - SafeDataType
+    /// * `content_type` - SafeContentType
+    /// * `path` - must already be percent-encoded if Some. leading '/' optional.
+    /// * `xorurl_sub_names` - sub_names. ignored if nrs_host is present.
+    /// * `query_string` - must already be percent-encoded, without ? separator
+    /// * `fragment` - url fragment, without # separator
+    /// * `content_version` - overrides value of "?v" in query-string if not None.
     pub fn new(
         xorname: XorName,
+        nrs_host: Option<&str>,
```
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:312]: sure.

File: safe-api/src/api/app/xorurl.rs [hunk 7]
```diff
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
...
+    /// returns XorName
     pub fn xorname(&self) -> XorName {
         self.xorname
     }
 
+    /// returns 'host' portion of xorurl using the
+    /// default xorurl encoding.
+    ///
+    /// For a different encoding, see host_to_base()
+    pub fn xorurl_host(&self) -> String {
```
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:584]: maybe xorurl_name

File: safe-api/src/api/app/xorurl.rs [hunk 8]
```diff
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
...
+            self.nrs_host.clone()
+        } else {
+            self.xorurl_host()
+        }
+    }
+
+    /// returns top-level-domain of host field.
+    ///
+    /// eg: my.sub.name --> name
+    pub fn tld(&self) -> String {
```
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:610]: perhaps just "name".

[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:610]: hmm, or maybe top_name(), to differentiate from name.

File: safe-api/src/api/app/xorurl.rs [hunk 9]
```diff
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
...
+    pub fn fragment_with_separator(&self) -> String {
+        if self.fragment.is_empty() {
+            "".to_string()
+        } else {
+            format!("#{}", self.fragment)
+        }
+    }
+
+    /// returns true if an NrsUrl, false if an XorUrl
+    pub fn is_nrs(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:825]: could be.  I also considered that maybe it should be like safeurl_type() and return an enumeration { nrsurl, xorurl }.   In case we ever need/want to add another type.

File: safe-api/src/api/app/xorurl.rs (hunk 4 above)
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:726]: yes.  without it:

```
error[E0502]: cannot borrow `url` as immutable because it is also borrowed as mutable
   --> safe-api/src/api/app/xorurl.rs:728:29
    |
704 |         let mut pairs = url.query_pairs_mut();
    |                         --- mutable borrow occurs here
...
728 |         self.query_string = url.query().unwrap_or("").to_string();
    |                             ^^^ immutable borrow occurs here
...
735 |     }
    |     - mutable borrow might be used here, when `pairs` is dropped and runs the destructor for type `url::form_urlencoded::Serializer<'_, url::UrlQuery<'_>>`
```
As you can see, that was the borrow checker's recommendation.  I'd be happy to remove it if you come up with an alternative that doesn't require the drop.

Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:726]: anyway, I can comment why it's there.

File: safe-api/src/api/app/xorurl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:610]: that's effectively the NRS public_name / nrs_name , isn't it?

[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:610]: Here's how I'm presently thinking about it:

```
    pub name: String,            // "a.b.name" in "a.b.name"
    pub sub_names: Vec<String>,  // "a.b"      in "a.b.name"
    pub top_name: String,        // "name"     in "a.b.name"
```

File: safe-api/src/api/app/xorurl.rs (hunk 3 above)
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:1265]: ok

File: safe-api/src/api/app/xorurl.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:825]: ok, I renamed is_nrs to is_nrsurl() and also added a SafeUrlType enum, and ::safeurl_type() getter:
```
    /// returns type of this SafeUrl.
    /// 
    /// for type of the linked content, see
    ///   ::content_type()
    pub fn safeurl_type(&self) -> SafeUrlType {
        if self.is_nrsurl() { 
            SafeUrlType::NrsUrl
        } else {
            SafeUrlType::XorUrl
        }
    }
```

File: safe-api/src/api/app/xorurl.rs (hunk 4 above)
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:726]: This should do it:
```
let mut pairs = url::form_urlencoded::Serializer::new(String::new());
let url = Self::query_string_to_url(&self.query_string)?;
let mut set_key = false;
for (k, v) in url.query_pairs() {
    if k == key {
        // note: this will consolidate multiple ?k= into just one.
        if let Some(v) = val {
            if !set_key {
                pairs.append_pair(key, v);
                set_key = true;
            }
        }
    } else {
        pairs.append_pair(&k, &v);
    }
}
if !set_key {
    if let Some(v) = val {
        pairs.append_pair(key, v);
    }
}

self.query_string = pairs.finish();
debug!("Set query_string: {}", self.query_string);
```

File: safe-api/src/api/app/xorurl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:610]: not sure if that matches and/or conflicts with how we've been calling it in NRS RFC, if not then fine.

File: safe-api/src/api/app/xorurl.rs (hunk 1 above)
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:843]: So I just reviewed again and the *only* place that can generate an error within this func is inside host_to_base() where it does a check that MediaType is not unknown.  This same/similar check is performed in ::new(), and there is no content_type setter.  In other words, if an error occurs, there is a bug in the implementation.  I reasoned that in this situation, it is acceptable to log warning and return an empty string.   The alternatives would seem to be:
 *  returning the error to caller, which we know/believe should never actually hit the error case.  
 * getting rid of the check in name_to_base(), since it is redundant with check in ::new().

I think I would prefer the latter, adding a comment that the validation is performed in new()

what do you think?

in ::host_to_base()
```
            SafeContentType::MediaType(media_type) => match MEDIA_TYPE_CODES.get(media_type) {
                Some(media_type_code) => *media_type_code,
                None => {
                    return Err(Error::Unexpected(format!(
                        "Failed to encode Media-type '{}'",
                        media_type
                    )))
                }
            },
```

in ::new()
```
        if let SafeContentType::MediaType(ref media_type) = content_type {
            if !Self::is_media_type_supported(media_type) {
                return Err(Error::InvalidMediaType(format!(
                        "Media-type '{}' not supported. You can use 'SafeContentType::Raw' as the 'content_type' for this type of content",
                        media_type
                    )));
            }
        }

```

Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:843]: Yes, I agree, the latter is the best option and what is usually done, any setter or constructor should take care of validations and return corresponding errors, so we can then have all type of getters (like to_string() or to_base()) to return a value and not a Result<>. Note we also are using the word host in that function that we need to rename/remove.

> it is acceptable to log warning and return an empty string
I think this is never acceptable, you are not handling error accordingly and hiding it. If you know it will never cause an error should not return a Result from a function, and it should always return a valid value. In this case you are forced to unwrap due to host_to_base() returning a Result<> which means it could fail, so you cannot ignore an error from that result/call, anyone making changes to the host_to_base() function may return a valid error as he/she sees the function returns a Result which means it's expected to return errors, but we won't be handling them correctly. Hiding "unexpected" errors like that is giving up the compilation check (which is why Result exists I believe), so you'll be having a bug sooner or later due to it.

File: safe-api/src/api/app/xorurl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:610]: It seems that property names never really matched [rfc-0053](https://github.com/maidsafe/rfcs/blob/master/text/0053-xor-urls/0053-xor-urls.md).  But if that's our goal (and it seems a reasonable one) we should probably do something like:
```
pub(crate) struct SafeUrlParts {
...
    pub name: String,               // "a.b.name" in "a.b.name"
    pub public_name: String,        // "name"     in "a.b.name"
    pub service: Vec<String>,       // "a.b"      in "a.b.name"
...
}

pub struct SafeUrl {
...
    xorname_uri: XorName,          // was xorname. applies to nrsurl and xorurl
    public_name_uri: String,       // was nrs_name.  // "a.b.name" in "a.b.name"
    public_name: String,           // new.              "name"     in "a.b.name"
    service: Vec<String>,          // was sub_names. // "a.b"      in "a.b.name"
...
}
```
and rename getters/setters to match.  I find these names more verbose and probably less descriptive than xorname, nrs_name, sub_name, but there is def something to be said for matching the rfc.  what do you think?


Also, rfc-0053 does not allow sub-names (service) for xor-uri (immutable-data-uri | mutable-data-uri), so we are out of spec on that.  As stated elsewhere, I'd like to drop that support, in a follow-up PR.

For quick reference:
```
<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
<immutable-data-uri>  := <cid> <query-fragment>
<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
```

[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:610]: 'service' is the old way we had but we moved away from it, so it sounds like the RFC is outdated in that regard it seems.
The sub-names for xorurls were not considered in the RFC I think, as I mentioned that was something you get for free based on the way we decode the public name and subnames. So you'll have to add the restriction if you want that from being resolved.

[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:610]: It'll be difficult to find exactly where we started using subname and forgetting about service name, but this was one of the main threads of discussions: https://forum.safedev.org/t/rfc-public-name-system-resolution-and-rdf/2185

[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:610]: Not sure if you were aware of this other RFC instead of XorUrl one (which is older): https://github.com/maidsafe/rfcs/blob/master/text/0052-RDF-for-public-name-resolution/0052-RDF-for-public-name-resolution.md and I see we are explicitly making use of public name and subnames only

File: safe-api/src/api/app/xorurl.rs (hunk 1 above)
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:843]: Looking at this MediaType code more closely, I see that the problem is that MediaType accepts an arbitrary String, eg "text/html", but then host_to_base() tries to look it up in a list of supported media types, in order to convert to a u16.  This lookup can return None, so there is no way to not handle the lookup failure.  I can think of a few possibilities:

1. We perform the lookup in ::new(), which is being done anyway, and save the u16 in the SafeUrl struct, else throw error.   only downside is that this requires extra space in the struct.
2. We ignore the possibility of error, because we know it is validated elsewhere. 
3. ::to_xorurl_string() must return a Result.

So there are tradeoffs with every possibility.  I tend to favor #1.  An extra u16 in the struct doesn't seem that bad.

Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs:843]: it sounds good to me, you will be having a u16 but you will be getting rid of the string you are storing now in the internal content_type field.

Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs:843]: ok, I implemented #1.  works fine.

> it sounds good to me, you will be having a u16 but you will be getting rid of the string you are storing now in the internal content_type field.

Unfortunately I could not remove the existing content_type field because it is actually a SafeContentType and is returned as such from ::content_type().  One might think it could be constructed from the u16, but in practice, SafeContentType::from_u16() can throw an error if input u16 is invalid.  (aside: looking at the ::from_value() routine, it doesn't actually support MediaType at all, so is incomplete.)  So the only way to keep content_type() API as it is, and remove result from to_xorurl_string() without ignoring any errors is to store both the SafeContentType and correesponding/validated u16 in the struct, which is a little gross, but works.

**Comment by dan-da** [comment]
I pushed latest commit in case anyone wants to take a look.  I did some renaming according to this scheme:

```
    pub name: String,            // "a.b.name" in "a.b.name"
    pub sub_names: Vec<String>,  // "a.b"      in "a.b.name"
    pub top_name: String,        // "name"     in "a.b.name"
```
but now its kind of an open question how well it should match up with rfc-0053 and/or rfc-0052.

Also, I added the content_type_u16 field, allowing to_xorurl_string() to return a String without ignoring any error.

I don't expect this to be merged yet.  More of a work-in-progress commit.

File: safe-api/src/api/app/xorurl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dan-da) [reply safe-api/src/api/app/xorurl.rs]: thx for links.  I'll read up on these discussions a bit.  If we don't go with the rfc naming, then I think it should be updated or superceded to match code/reality.

[the remark this answers is not in the document]
Comment (bochaco) [reply safe-api/src/api/app/xorurl.rs]: RFC52 is probably the most up to date with discussions that happened on the forum, as that was finalised after the xorurl one if i remember correctly, it would be good to hear @joshuef input here too as he was leading that RFC
